x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotRepository.java

```diff
@@ -0,0 +1,253 @@
+            try (HardlinkCopyDirectoryWrapper wrapper = new HardlinkCopyDirectoryWrapper(FSDirectory.open(restoreSourceCopy))) {
+                Lucene.cleanLuceneIndex(wrapper);
+                SegmentInfos segmentInfos = shard.store().readLastCommittedSegmentsInfo();
+                for (String file : segmentInfos.files(true)) {
+                    wrapper.copyFrom(shard.store().directory(), file, file, IOContext.DEFAULT);
+                }
+            }
+            Lucene.cleanLuceneIndex(shard.store().directory()); // wipe the old index
+            shard.store().createEmpty();
+        } catch (IOException ex) {
+            // why on earth does this super method not declare IOException
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    @Override
+    public void applyPostRestoreOps(IndexShard shard) throws IOException {
+        ShardPath shardPath = shard.shardPath();
+        Path restoreSourceCopy = shardPath.getDataPath().resolve(RESTORE_DIR_NAME);
+        RecoveryState.Translog state = shard.recoveryState().getTranslog();
+        assert state.totalOperations() == 0 : "translog state should have 0 total ops but got: " + state.totalOperations();
+        state.reset();
+        long sequenceNumber = 0;
+        try (FSDirectory dir = FSDirectory.open(restoreSourceCopy)) {
+            try (IndexReader reader = DirectoryReader.open(dir)) {
+                state.totalOperationsOnStart(reader.numDocs());
+                state.totalOperations(reader.numDocs());
+                long primaryTerm = shard.getPendingPrimaryTerm();
+                FieldsVisitor rootFieldsVisitor = new FieldsVisitor(true);
+                for (LeafReaderContext ctx : reader.leaves()) {
+                    LeafReader leafReader = ctx.reader();
+                    Bits liveDocs = leafReader.getLiveDocs();
+                    // TODO: we could do this in parallel per segment here or even per docID
+                    // there is a lot of room for doing this multi-threaded but not for the first iteration
+                    for (int i = 0; i < leafReader.maxDoc(); i++) {
+                        if (liveDocs == null || liveDocs.get(i)) {
+                            rootFieldsVisitor.reset();
+                            leafReader.document(i, rootFieldsVisitor);
+                            rootFieldsVisitor.postProcess(shard.mapperService());
+                            Uid uid = rootFieldsVisitor.uid();
+                            BytesReference source = rootFieldsVisitor.source();
+                            if (source != null) { // nested fields don't have source. in this case we should be fine.
+                                // TODO we should have a dedicated origin for this LOCAL_TRANSLOG_RECOVERY is misleading.
+                                Engine.Result result = shard.applyTranslogOperation(new Translog.Index(uid.type(), uid.id(),
```

Comment:
this is a pretty terrible abuse of the Translog phase. we should rename it to Ops phase or so. I spoke with @bleskes about this. This is a shortcut and needs some discussion. also with @ywelsch